src/algorithms/blocked_gibbs_algorithm.cc

```diff
@@ -1,22 +1,55 @@
 #include "blocked_gibbs_algorithm.h"
 
+#include <stan/math/prim/fun.hpp>
+
 #include "hierarchy_id.pb.h"
 #include "mixing_id.pb.h"
 #include "src/hierarchies/base_hierarchy.h"
 #include "src/mixings/base_mixing.h"
+#include "src/utils/distributions.h"
+#include "src/utils/rng.h"
 
 void BlockedGibbsAlgorithm::print_startup_message() const {
   std::string msg = "Running BlockedGibbs algorithm with " +
                     bayesmix::HierarchyId_Name(unique_values[0]->get_id()) +
                     " hierarchies, " +
-                    bayesmix::MixingId_Name(mixing->get_id()) + " mixing...";
+                    bayesmix::MixingId_Name(cond_mixing->get_id()) +
+                    " mixing...";
   std::cout << msg << std::endl;
 }
 
 void BlockedGibbsAlgorithm::sample_allocations() {
-  return;  // TODO
+  auto &rng = bayesmix::Rng::Instance().get();
+  unsigned int num_components = cond_mixing->get_num_components();
+  for (int i = 0; i < data.rows(); i++) {
+    // Compute weights
+    Eigen::VectorXd logprobas =
+        cond_mixing->get_weights(true, false, mix_covariates.row(i));
+    for (int j = 0; j < num_components; j++) {
+      logprobas(j) +=
+          unique_values[j]->like_lpdf(data.row(j), hier_covariates.row(j));
```

Comment:
this should be data.row(i), this is likely to explain the bad ARIs

Reply:
🤦 FML. Thanks a bunch! Hope this fixes all issues.

Reply:
Next homework is to think some unit-tests for these situations :+1:

Reply:
I know right. I caught a bug a few days ago which involved a proto writing error (I still don't know why that was an error though), and it was the ONLY read-write functionality I forgot to unit-test. These tests just don't come naturally to mind.